text/0000-forget-marker-trait.md

```diff
@@ -0,0 +1,790 @@
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+All types were assumed to be `!Forget` in Rust's early days, and then it was changed in hurry. They flow naturally out of Rust's type system, do not clash with any preexisting concepts that do not directly involve forgetting and are used pleasantly and intuitively, modulo migration. With the `async` built around `Future` trait it became apparent that language directly lacks this feature. Being simple and non-disturbing, it's hard to find something that would fit that purpose better.
+
+We can do nothing, but use cases just keep piling up.
+
+The author of https://zetanumbers.github.io/book/myosotis.html is working on another approach to that problem, but it is not public yet.
+
+# Prior Art
+[prior-art]: #prior-art
+
+- https://github.com/rust-lang/rfcs/pull/1084#issuecomment-96875651
+- https://github.com/rust-lang/rfcs/pull/1094
+- https://internals.rust-lang.org/t/forgetting-futures-with-borrowed-data/10824
+- https://github.com/aturon/rfcs/blob/scoped-take-2/text/0000-scoped-take-2.md
+- https://without.boats/blog/the-scoped-task-trilemma
+- https://without.boats/blog/asynchronous-clean-up/
+- https://zetanumbers.github.io/book/myosotis.html: an independent exploration of the same problem space with similar, but subtly different, conclusions.
+- https://hackmd.io/@wg-async/S1Q6Leam0: a design meeting regarding the previous post
+
+## Leakpocalypse
+[leakpocalypse-prior-art]: #leakpocalypse-prior-art
+
+- https://github.com/rust-lang/rfcs/pull/1066
+- https://github.com/rust-lang/rust/issues/24292
+- https://cglab.ca/~abeinges/blah/everyone-poops/
+- https://github.com/rust-lang/rfcs/pull/1085
+- https://doc.rust-lang.org/std/thread/fn.scope.html
+- https://smallcultfollowing.com/babysteps/blog/2015/04/29/on-reference-counting-and-leaks/
+
+## Usage of the pattern
+[usage-prior-art]: #usage-prior-art
+
+- https://doc.rust-lang.org/std/thread/struct.Builder.html#method.spawn_unchecked
+- https://docs.rs/async-task/latest/async_task/fn.spawn_unchecked.html
+- https://blog.japaric.io/safe-dma/
+- https://docs.rs/async_nursery/latest/async_nursery/
+- https://without.boats/blog/the-scoped-task-trilemma/
+- https://docs.rs/async-scoped/latest/async_scoped/struct.Scope.html#method.scope_and_collect
+
+## Miscellaneous
+[misc-prior-art]: #misc-prior-art
+
+- https://github.com/rust-lang/rfcs/issues/1111
+- https://tmandry.gitlab.io/blog/posts/2023-03-01-scoped-tasks/
+
+## MustMove types
+[must-move-prior-art]: #must-move-prior-art
+
+- https://faultlore.com/blah/linear-rust/
+- https://blog.yoshuawuyts.com/linear-types-one-pager/
+- https://smallcultfollowing.com/babysteps/blog/2023/03/16/must-move-types/
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- [ ] Maybe the name `Forget` is misleading, as its core is around the `unsafe` guarantee of borrowed resources and the destructor. `IndirectPin`?
```

Comment:
I'd rather not mix it with `Pin`. `Pin` is an advanced concept needed only for certain niches, so for many users it may not clarify the meaning, but rather connect it to something they're unfamiliar with. There are also proposals for integrating `Pin` more into Rust's syntax, so the `Pin` itself may become even less visible in the future.

I assume that `Forget` may appear more often in the API surface. It may eventually be easier to teach `Pin` in terms of `!Forget` than the other way.